[EIP-4762]: clarify EIP using python for spec, add atomicity, warm costs and 7702 (#9780)
ethereum/EIPs: EIPS/eip-4762.md

## Motivation

> The introduction of Stateless Ethereum requires fundamental changes, and this EIP introduces changes to the gas schedule, reflectiong the new storage model, which seeks to disincentivize witness growth as well as state growth.

Reviewer: ```suggestion
The introduction of Stateless Ethereum requires fundamental changes, and this EIP introduces changes to the gas schedule, reflecting the new storage model, which seeks to disincentivize state and witness growth.
```

### Access events

> Whenever the state is read or written to, one or more of the access events of the form`(address, sub_key, leaf_key)` take place, determining what data is being accessed. We define access events as follows:

Reviewer: ```suggestion
Whenever the state is read or written to, one or more of the access events of the form `(address, sub_key, leaf_key)` take place, determining what data is being accessed. We define access events as follows:
```

```python3
def add_access_event(address: bytes20, sub_key: bytes31, leaf_key: byte, is_write: bool):
  event = (address, sub_key_leaf_key)
  if is_write:
    tx.access_events.writes.append(event)
  else:

>     tx.acess_events.reads.append(event)

Reviewer 2: what is `tx`? I would maybe just remove it to avoid confusion

An event will be triggered:

 * Every time a group is accessed
 * Every time a single leaf is accessed

The model makes a distinction between read and write events. As we will see later, gas costs are associated to these events.

#### Atomicity

The calls to `charge_gas()` are here to delimit the atomic boundaries of an operation. All the events emitted between two consecutive calls to `charge_gas` are charged together. If there isn't enough gas to pay for all events, none of the associated values are added to the witness.

#### Access events for account headers

```python3
# Emit all access events given an instruction. It is meant to be called before
# the gas charging method for a given instruction is called, as these events
# are subsequently used to process witness-related gas costs.
def emit_access_events(opcode: int, stack: Stack, contract: Contract, new_contract: Optional[Contract]):
  match opcode:
    case OP_CALL | OP_DELEGATECALL | OP_STATICCALL | OP_CALLCODE | OP_SELFDESTRUCT:
      address = stack.peek(1)

      # XXX the previous version of the spec says OP_CALLCODE is concerned, but the geth

>       # code says it isn't.

Reviewer: Inconsistent variable name (sometimes `opcode` sometimes `op` for the same variable)

>       # code says it isn't.

Reviewer 2: We should fix that CALLCODE case and decided which do we want to go with. Besu also ignores the value bearing part as the caller transfers value to itself

>       if value_bearing and (op == OP_CALL or op == OP_SELFDESTRUCT or op == OP_CALLCODE):

Reviewer 2: Can't see the `value_bearing` variable declared

>         add_write_access_event(contract.address, 0, BASIC_DATA_LEAF_KEY)

Reviewer 2: So a read always precedes a write, but that read could be of a warm account. I don't see any check on whether the read part was on a warm account or not.

        add_write_access_event(address, 0, BASIC_DATA_LEAF_KEY)

        # If the operation triggers an account creation, charge the
        # full cost of account creation
        if not state.has_account(address):

>           add_write_access_event(address, 0, CODEHASH_LEAF_KEY)

Reviewer 2: This is certainly a SET kind of write because the account didn't existed. You need to add a distinction for the 2 different writes (SET and RESET).

      else:
        if is_precompile(address) or is_system_contract(address):

>           # XXX charge warm costs

Reviewer: Is this a todo?

Author: yes

        else:
          add_read_access_event(address, 0, BASIC_DATA_LEAF_KEY)
       
      add_read_access_event(contract.address, 0, BASIC_DATA_LEAF_KEY)

    case OP_EXTCODECOPY | OP_EXTCODESIZE:
      address = stack.peek(0)

>       add_read_access_event(address, 0, BASIC_DATA_LEAF_KEY)

Reviewer 2: missing precompile / system contract check to charge warm costs

    case OP_BALANCE:
      address = stack.peek(0)

>       add_read_access_event(address, 0, BASIC_DATA_LEAF_KEY)

Reviewer 2: missing precompile / system contract check to charge warm costs

    case OP_CODECOPY:
       if not contract.is_deployment:

>          add_read_access_event(contract.address, 0, BASIC_DATA_LEAF_KEY)

Reviewer 2: I believe this has no effect because current account should already be in the witness

    case OP_EXTCODEHASH:
      address = stack.peek(0)
      if not (is_precompile(address) or is_system_contract(address)):
        add_read_access_event(address, 0, CODEHASH_LEAF_KEY)

    case OP_SLOAD:
      address = stack.peek(0)

>       key = stack.peek(1)

Reviewer: I think SLOAD uses the current contract address, not a stack argument. Then SLOAD's key is at stack position 0, not 1.

>       read_access_for_storage(address, key)

Reviewer 2: Charge warm gas

    case OP_SSTORE:
      key = stack.peek(0)

>       write_access_for_storage(contract.address, key)

Reviewer 2: function does not exist in the document

      key = stack.peek(1)

>       write_access_for_storage(address, key)

Reviewer: Same as above, contractAddress instead of stack.peek(0), and stack.peek(0) as key

      # Explitly specify that the witness costs for contract
      # creation are not charged as part of the instruction,

>       # but as part of the creation execution it triggers.

Reviewer 3: How does this work if CREATE2 tries to deploy a contract at an address which already has a contract (thus deployed before by CREATE2 with the same inputs)

Reviewer 3: (should this call `emit_access_events_for_contract_creation_account_check`?)

      # creation are not charged as part of the instruction,
      # but as part of the creation execution it triggers.
      # This means, among other things, that the subtraction
      # of 1/64th of the gas is done before.

  charge_gas()

# This function is to be called right when it is known that the
# account will be created.
def emit_access_events_for_contract_creation_init(address: Address):
  add_write_access_event(address, 0, BASIC_DATA_LEAF_KEY)
  add_write_access_event(address, 0, CODEHASH_LEAF_KEY)
  charge_gas()

def emit_access_events_for_contract_creation_account_check(address: Address):

>   add_read_access_event(new_contract.address, 0, BASIC_DATA_LEAF_KEY)
>   add_read_access_event(new_contract.address, 0, CODEHASH_LEAF_KEY)

Reviewer: ```suggestion
  add_read_access_event(address, 0, BASIC_DATA_LEAF_KEY)
  add_read_access_event(address, 0, CODEHASH_LEAF_KEY)
```

  charge_gas()
  
def read_access_for_storage(contract_address: bytes20, storage_key: byte32):

>   index, suffix = get_storage_slot_key(contract_address, storage_key)

Reviewer: ```suggestion
  index, suffix = get_storage_slot_tree_keys(contract_address, storage_key)
```

```

##### Access events for delegation

> When `contract_address` is that of a delegating EoA, process the following events:

Reviewer 3: Clarify that "delegating EoA" is thus in the context of EIP-7702 (right?)

 - For an `EXTCODEHASH`:

```
(delegated_address, 0, CODEHASH_LEAF_KEY)
```

>  - For an `EXTCODESIZE`, or any other `*CALL`, `CALLCODE`:

Reviewer 3: What about `EXTCODECOPY`?

 - For an `EXTCODESIZE`, or any other `*CALL`, `CALLCODE`:

```

> (delegated_address, 0, BASIC_DATA_LEAF_KEY)

Reviewer: Should we specify whether or not these event accesses are write or read? The previous EIP mentioned that these are write accesses

Author: that part still needs to be pythonized but yes we will

```

#### Access events for code

```python3
def interpreter_loop(code: bytes, contract_address: bytes20):
  pc = 0

  # ...

  while True:
    if pc < len(code):
      emit_code_access_events(pc, code, contract_address)
      # ... rest of the interpreter loop

# Executed when decoding an instruction when running the interpreter loop

> def emit_code_access_events(pc: int, code: bytes, contract_address: Address):

Reviewer 2: I didn't see this anywhere within this function - should you not check if you are running a contracts init code too?

  # XXX I don't think that this is true on mainnet, even before the atomic stuff!

>   if pc >= len(code):

Reviewer 3: Is the basic account data loaded if we load the data from the corresponding `BASIC_DATA_LEAF_KEY`? Because then we know `len(code)`. If we know this there is no need to include a chunk which is past the code size (since we already know the code size). Same reasoning for JUMP/JUMPIs since that would jump to a non-JUMPDEST and should thus OOG without needing to include the code chunk. If `len(code)` is unknown then somehow a proof should be included that the chunk is empty.

Reviewer 3: What about a PUSH opcode which for instance is a PUSH32 which sits at pc 16 and only has 8 bits filled? (code size is thus 24) How would this work? Mainnet would assume that the empty bytes are 0s and push that (so right-align the pushed value) and would then thus run out-of-bounds (STOP opcode) . See YP (pic) for right-padding these partially out-of-bounds PUSHs.
![image](https://github.com/user-attachments/assets/1ae0ecae-8346-43de-bdfc-6de1d9a27329)

  # XXX I don't think that this is true on mainnet, even before the atomic stuff!
  if pc >= len(code):
    return

  opcode = code[pc]
  chunk = pc // CHUNK_SIZE

  add_read_access_event(contract_addresss, (chunk + CODE_OFFSET) // 256, (chunk + CODE_OFFSET) % 256)
  charge_gas() # Note that the witness gas for instruction access is charged immediately,
               # independently of access costs for its immediates.

  # add access events for immediates, if the instruction has them
  match opcode:
    case OP_JUMP:
      dest = stack.peek(0)
      # The destination of a jump is not considered to be accessed if it is beyond the code
      if dest < len(code):
        # The destination of a `JUMP` is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata

>         add_read_access_event(dest)

Reviewer: To fit the signature of the method defined at the top, this is missing two params: subkey, and leaf_key

>     case OP_JUMPI:

Reviewer 3: ```suggestion
    case OP_JUMPI | OP_JUMP:
```

      dest = stack.peek(0)

>       cond = stack.peek(1)

Reviewer 3: General question about `stack.peek`, what if it tries to peek "outside" stack height? Does it (still) trigger out-of-gas?

      # The destination of a jump is not considered to be accessed if it is beyond the code
      if dest < len(code):
        dest_chunk = dest // CHUNK_SIZE
        # The destination of a positively evaluated `JUMPI` is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata
        # The destination of a `JUMPI` is not considered to be accessed if the jump conditional is `false`

>         if opcode == OP_JUMP or cond:

Reviewer 2: how can `opcode == OP_JUMP` here?

Author: it can't, it's leftover from a refactor, thanks for catching that.

        # The destination of a `JUMPI` is not considered to be accessed if the jump conditional is `false`
        if op == OP_JUMP or cond:
          add_read_access_event(contract_address, (dest_chunk + CODE_OFFSET) // 256, (dest_chunk + CODE_OFFSET) % 256)

    case OP_PUSH1 | OP_PUSH2 | ... | OP_PUSH32:
      nbytes = opcode - OP_PUSH0 # determine the number of extra bytes to touch
      for i in range(nbytes):
        immediate_chunk = (pc+i) // CHUNK_SIZE
        if immediate_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:
          add_read_access_event(contract_address, (immediate_chunk + CODE_OFFSET) // 256, (immediate_chunk + CODE_OFFSET) % 256)

    case OP_CODECOPY | OP_EXTCODECOPY:
      addr = contract_address
      size = 0

>       offset = i0

Reviewer: what's i0?

Author: a typo :innocent:

        offset = stack.peek(2)
        size = stack.peek(3)
      else:
        offset = stack.peek(1)
        size = stack.peek(2)

      for i in range(size):
        copied_byte_chunk = (offset+i) // CHUNK_SIZE 

>         if copied_byte_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:

Reviewer 3: I think this is easier (removes the `- 1`)
```suggestion
        if copied_byte_chunk < (len(code)+CHUNK_SIZE) // CHUNK_SIZE:
```

`code` refers to locally loaded code which will not work for EXTCODECOPY on non-local addresses

          add_read_access_event(contract_address, (copied_byte_chunk + CODE_OFFSET) // 256, (copied_byte_chunk + CODE_OFFSET) % 256)

    case OP_CODESIZE | OP_EXTCODESIZE | OP_EXTCODEHASH:
      # None of these operations access any chunk, it is

>       # explicitly documented here as a no-op.

Reviewer 2: why are these JUMP, JUMPI, PUSHN, CODECOPY, EXTCODECOPY declared directly inside the interpreter loop and not above in L75 like any other ones?

  charge_gas()
```

Note: since no access list existed for code up until this EIP, note that no warm costs are charged for code accesses.

> When a contract is created, access chunks `0 ... (len(code)+30)//31`:

Reviewer 3: ```suggestion
When a contract is created, access chunks `0 ... (len(code)+CHUNK_SIZE-1)//CHUNK_SIZE`:
```

(Maybe note this is a integer ceil operation also)

When a contract is created, access chunks `0 ... (len(code)+30)//31`:

```python3
def emit_code_write_access_events(address: Address, code: bytes):

>   last_chunk_num = (len(code) + CHUNK_SIZE - 1) // CHUNK_SIZE

Reviewer: ```suggestion
    # Calculate the number of code chunks required, rounding up for any partial final chunk.
    # This ensures that all code bytes are covered, even if the code size is not a multiple of CHUNK_SIZE.
  last_chunk_num = (len(code) + CHUNK_SIZE - 1) // CHUNK_SIZE
```
This is a caveat that is easy to miss and can lead to hard to spot bugs, imo worth adding a comment

  for chunk_num in range(last_chunk_num):

>     add_write_access_event(contract_address, (chunk_num + CODE_OFFSET) // 256, (chunk_num + CODE_OFFSET) % 256)

Reviewer: ```suggestion
    add_write_access_event(address, (chunk_num + CODE_OFFSET) // 256, (chunk_num + CODE_OFFSET) % 256)
```

  charge_gas()
```

### Transaction

For a transaction, make these access events:

```python3
def emit_tx_prologue_events(from: Address, to: Address, value: int):
  add_write_access_event(from, 0, BASIC_DATA_LEAF_KEY) # Note this is the only write, because of the nonce

>   add_read_access_event(from, 0, CODEHASH_LEAF_KEY)

Reviewer 3: Why do we read this key? Is this because of https://eips.ethereum.org/EIPS/eip-3607 ?

  add_read_access_event(to, 0, BASIC_DATA_LEAF_KEY)
  add_read_access_event(to, 0, CODEHASH_LEAF_KEY)

  if value != 0:
    add_write_access_event(to, 0, BASIC_DATA_LEAF_KEY)

>     if not state.has_account(to)

Reviewer 2: should charge a write SET in this case for BASIC_DATA and CODEHASH if account does not exist yet

>   add_read_access_event(to, 0, CODEHASH_LEAF_KEY)

Reviewer 3: Should update to also read/write to accounts being delegated by 7702 txs

      add_write_access_event(to, 0, CODEHASH_LEAF_KEY)

  # Note that gas isn't charged, as it's considered covered by
  # the transaction's intrinsic gas.
```

For a contract-creating transaction, the emitted events are defined by:

```python
def contract_creation(address: Address, code: bytes):
  # ...

>   # Just after increating the caller's nonce

Reviewer: ```suggestion
  # Just after incrementing the caller's nonce
```
